### PR TITLE
[RFR] Update documentation about services autoconfigure & Symfony autowiring

### PR DESCRIPTION
**Pull request by vincentchalamon**
From https://github.com/api-platform/core/pull/1152

### CODE REVIEW

File: core/data-providers.md [hunk 1]
```diff
@@ -70,6 +70,16 @@ register and use this data provider. The optional attribute `priority` allows to
 data providers. The first data provider not throwing a `ApiPlatform\Core\Exception\ResourceClassNotSupportedException`
 will be used.
 
+**Note: Using Symfony 3.3 autowiring, service declaration is automatized. Update your autowiring configuration to add
```
Comment (teohhanhui) [inline core/data-providers.md]: automatized -> automated

File: core/extensions.md [hunk 2]
```diff
@@ -137,6 +137,16 @@ Thanks to the `api_platform.doctrine.orm.query_extension.collection` tag, API Pl
...
 Notice the priority level for the `api_platform.doctrine.orm.query_extension.collection` tag. When an extension implements the `ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryResultCollectionExtensionInterface` or the `ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryResultItemExtensionInterface` interface to return results by itself, any lower priority extension will not be executed. Because the pagination is enabled by default with a priority of 8, the priority of the `app.doctrine.orm.query_extension.current_user` service must be at least 9 to ensure its execution.
 
+**Note: Using Symfony 3.3 autowiring, service declaration is automated. Update your autowiring configuration to add
+`Doctrine\ORM\Extension` directory:**
+
+```yml
+# app/config/services.yml
+services:
+    App\:
+        resource: '%kernel.project_dir%/src/{Doctrine/ORM/Extension,Controller,Action,Command,EventSubscriber}'
```
Comment (teohhanhui) [inline core/extensions.md]: I really do not like that it's under `Doctrine/ORM/Extension`, IMO it should be `ApiPlatform/Doctrine/ORM/Extension` to avoid causing confusion that it's a Doctrine thing.

File: core/data-providers.md [hunk 3]
```diff
@@ -70,6 +70,16 @@ register and use this data provider. The optional attribute `priority` allows to
...
 will be used.
 
+**Note: Using Symfony 3.3 autowiring, service declaration is automated. Update your autowiring configuration to add
+`DataProvider` directory:**
+
+```yml
+# app/config/services.yml
+services:
+    App\:
+        resource: '%kernel.project_dir%/src/{DataProvider,Controller,Action,Command,EventSubscriber}'
```
Comment (teohhanhui) [inline core/data-providers.md]: Same here, I believe it should be `ApiPlatform/DataProvider`.

File: core/extensions.md (hunk 2 above)
Comment (dunglas) [inline core/extensions.md]: May I suggest to just use `Extension`. It's shorter (more in the spirit of Flex) and we know that it's an API Platform project.

Comment (teohhanhui) [inline core/extensions.md]: `Extension` is completely non-descriptive. And even API Platform itself could have different sub-namespaces with the name `Extension` in them.

But I kinda agree and disagree on the spirit of Flex argument. The way I see it, Flex is about reclaiming the project. It's no longer an **API Platform** project, but **your** project. So a descriptive name is definitely needed...

**Comment by vincentchalamon** [comment]
@teohhanhui Fixed. But you should open a PR to fix the documentation to move `Filters` to `ApiPlatform/Filters` directory

File: core/data-providers.md [hunk 4]
```diff
@@ -30,9 +30,9 @@ If no data is available, you should return an empty array.
 ```php
 <?php
 
-// src/AppBundle/DataProvider/BlogPostCollectionDataProvider.php
+// src/AppBundle/ApiPlatform/DataProvider/BlogPostCollectionDataProvider.php
```
Comment (dunglas) [inline core/data-providers.md]: Why this change?

Comment (vincentchalamon) [inline core/data-providers.md]: https://github.com/api-platform/docs/pull/214#pullrequestreview-41552803

File: core/data-providers.md (hunk 3 above)
Comment (dunglas) [inline core/data-providers.md]: I don't agree. It increases the perceived complexity. We should keep it as simple as possible. The user can introduce a deeper hierarchy if he has too, but in the doc, it should keep it simple stupid.

And as already stated, it's more in the spirit of Flex, that we will integrate in API Platform.

File: core/data-providers.md [hunk 5]
```diff
@@ -70,6 +70,16 @@ register and use this data provider. The optional attribute `priority` allows to
...
 will be used.
 
+**Note: Using Symfony 3.3 autowiring, service declaration is automated. Update your autowiring configuration to add
+`DataProvider` directory:**
+
+```yml
+# app/config/services.yml
+services:
+    App\:
+        resource: '%kernel.project_dir%/src/{ApiPlatform/DataProvider,Controller,Action,Command,EventSubscriber}'
```
Comment (teohhanhui) [inline core/data-providers.md]: There is a mismatch here (no `AppBundle`).

Comment (vincentchalamon) [inline core/data-providers.md]: In Symfony 3.3, it's not recommended anymore to use a `AppBundle`

Comment (dunglas) [inline core/data-providers.md]: Only if you use Flex. If you use the standard edition, it's still the best practice (yes, it's a bit complicated right now). As API Platform uses the standard edition directory structure for now, I suggest to use: `src/AppBundle/DataProvider` for (without mentioning `ApiPlatform`).

When we'll switch to Flex (and we will do it), we'll update the whole documentation repository in one time to match the new directory structure.

Comment (vincentchalamon) [inline core/data-providers.md]: Fixed
@dunglas can you update your review please?

File: core/extensions.md [hunk 6]
```diff
@@ -137,6 +137,16 @@ Thanks to the `api_platform.doctrine.orm.query_extension.collection` tag, API Pl
...
 Notice the priority level for the `api_platform.doctrine.orm.query_extension.collection` tag. When an extension implements the `ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryResultCollectionExtensionInterface` or the `ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryResultItemExtensionInterface` interface to return results by itself, any lower priority extension will not be executed. Because the pagination is enabled by default with a priority of 8, the priority of the `app.doctrine.orm.query_extension.current_user` service must be at least 9 to ensure its execution.
 
+**Note: Using Symfony 3.3 autowiring, service declaration is automated. Update your autowiring configuration to add
+`Doctrine\ORM\Extension` directory:**
+
+```yml
+# app/config/services.yml
+services:
+    AppBundle\:
+        resource: '%kernel.project_dir%/src/AppBundle/{Doctrine/ORM/Extension,Controller,Action,Command,EventSubscriber}'
```
Comment (teohhanhui) [inline core/extensions.md]: -1 for this. `Doctrine\ORM` namespace is owned by the Doctrine project. I have no doubt that this will confuse the heck out of many users.

Comment (teohhanhui) [inline core/extensions.md]: Perhaps we can call it `DataProviderExtension`, that's what it really is.

Comment (vincentchalamon) [inline core/extensions.md]: I kept the same structure than in the doc: https://api-platform.com/docs/core/extensions#example
If it needs to be updated, I think it should be done in another PR.
